src/commands/watch.rs

```diff
@@ -78,182 +73,177 @@ pub struct WatchCommand {
+
+        let spin_bin = std::env::current_exe()?;
+        let manifest_file = spin_common::paths::resolve_manifest_file_path(&self.app_source)?;
+        let manifest_dir = parent_dir(&manifest_file)?;
+        let manifest = spin_loader::local::raw_manifest_from_file(&manifest_file)
+            .await?
+            .into_v1();
 
-        // Start watching
-        let runtime = Watchexec::new(init_config, runtime_config.clone())?;
-        runtime
-            .send_event(Event::default(), Priority::Urgent)
-            .await?;
-        runtime.main().await??;
-        Ok(())
-    }
+        // Set up the event processors (but don't start them yet).
+        // We create the build-related objects even if skip_build is on - the cost is insignificant,
+        // and it saves having to check options all over the place.
+
+        let (artifact_tx, artifact_rx) = tokio::sync::watch::channel(Uuid::new_v4());
+        let (pause_tx, pause_rx) = tokio::sync::mpsc::channel(1);
+        let (source_code_tx, source_code_rx) = tokio::sync::watch::channel(Uuid::new_v4());
+        let (stop_tx, stop_rx) = tokio::sync::watch::channel(Uuid::new_v4());
+
+        let mut buildifier = Buildifier {
+            spin_bin: spin_bin.clone(),
+            manifest: manifest_file.clone(),
+            clear_screen: self.clear,
+            watched_changes: source_code_rx,
+            uppificator_pauser: pause_tx,
+        };
 
-    fn generate_command(&self) -> String {
-        // The docs for `current_exe` warn that this may be insecure because it could be executed
-        // via hard-link. I think it should be fine as long as we aren't `setuid`ing this binary.
-        String::from(
-            std::env::current_exe()
-                .unwrap()
-                .to_str()
-                .expect("to find exe path"),
-        )
-    }
+        let mut uppificator = Uppificator {
+            spin_bin: spin_bin.clone(),
+            manifest: manifest_file.clone(),
+            up_args: self.up_args.clone(),
+            clear_screen: self.clear,
+            watched_changes: artifact_rx,
+            pause_feed: pause_rx,
+            stopper: stop_rx,
+        };
 
-    fn generate_arguments(
-        state: State,
-        up_args: Vec<String>,
-        manifest_path: String,
-    ) -> Vec<String> {
-        let mut spin_args = match state {
-            State::Building => vec![String::from("build")],
-            State::Running => vec![String::from("up")],
+        // Start `watchexec` tasks to monitor artifact and build files.
+
+        // BUG: Currently the filter is not dynamic - if the asset list in the manifest changes, this will
+        // trigger a reload but the old asset list remains effecitve. I think this was an existing issue, not
+        // sure if I've exacerbated it. We can use `Watchexec::reconfigure()` to fix this but
+        // then we might need a Reconfiguriser as well to monitor the manifest.
```

Comment:
Since it's not a regression, I'll defer it to a followup PR.  That minimises the complexity of this one and keeps it as a pure bug fix of the existing functionality, so that people can assess the impact of the re-read feature separately.

Reply:
Sounds good to me